src/components/bottom_bar/bottom_bar.ts

```diff
@@ -209,6 +240,88 @@ export class BottomBar extends Component<Props, SpreadsheetChildEnv> {
     this.sheetListRef.el.scrollTo({ top: 0, left: scroll, behavior: "smooth" });
   }
 
+  onSheetMouseDown(sheetId: UID, event: MouseEvent) {
+    if (event.button !== 0) return;
+    this.closeMenu();
+
+    const mouseX = event.clientX;
+
+    document.body.style.cursor = "move";
+    const visibleSheets = this.getVisibleSheets();
+    const sheetRects = this.getSheetItemRects();
+
+    const sheets = visibleSheets.map((sheet, index) => ({
+      id: sheet.id,
+      size: sheetRects[index].width,
+      position: sheetRects[index].x,
+    }));
+    this.dndHelper = new DOMDndHelper({
+      draggedItemId: sheetId,
+      mouseX,
+      items: sheets,
+      containerEl: this.sheetListRef.el!,
+      onChange: (newPositions) => {
+        this.sheetState.isDnd = true;
+        this.sheetState.sheetDndPositions = newPositions;
+      },
+      onCancel: () => this.stopDragging(),
+      onDragEnd: (sheetId: string, finalIndex: number) => this.onDragEnd(sheetId, finalIndex),
+    });
+  }
+
+  private onDragEnd(sheetId: string, finalIndex: number) {
+    const originalIndex = this.sheetState.sheetList.findIndex((sheet) => sheet.id === sheetId);
+    const delta = finalIndex - originalIndex;
+    if (sheetId && delta !== 0) {
+      this.env.model.dispatch("MOVE_SHEET", {
+        sheetId: sheetId,
+        delta: delta,
+      });
+    }
+    this.stopDragging();
+  }
+
+  getSheetStyle(sheetId: UID): string {
+    const style: CSSProperties = {};
+    if (this.sheetState.isDnd) {
+      style.position = "relative";
+      style.left = (this.sheetState.sheetDndPositions?.[sheetId] || 0) + "px";
+      style.transition = "left 0.5s";
+      style.cursor = "move";
+    }
+    if (this.isDragged(sheetId)) {
+      style.transition = "left 0s";
+      style["z-index"] = "1000";
+    }
+    return cssPropertiesToCss(style);
+  }
+
+  private stopDragging() {
+    document.body.style.cursor = "";
+    this.sheetState.sheetList = this.getVisibleSheets();
+    this.sheetState.isDnd = false;
+    this.sheetState.sheetDndPositions = undefined;
+    this.dndHelper = undefined;
+  }
+
+  private getSheetItemRects(): Rect[] {
+    return Array.from(document.querySelectorAll<HTMLElement>(`.o-sheet`))
```

Comment:
We should avoid querying elements from `document` when it's possible. It's not a good practice as it's not self-contained and the code which integrates o-spreadsheet could use `o-sheet` for something else.  Here you can use `bottomBarRef`